[로또] 1단계 미션 제출합니다.

owenoweny opened the pull request:
## 고민 사항

- 최대한 객체를 wrapping해보려 했습니다. 단순히 모든 것을 감싸면 복잡도가 너무 높아질 것 같다고 생각했는데, 적절한 지점을 찾는 것이 어렵다고 느꼈습니다. 
- 클래스 사이의 역할을 잘 분배하고 각각이 응집성 있게 구현하기 위해서 노력했습니다. 

위와 관련해서 피드백 주시면 감사드리겠습니다.

adeldel commented on src/main/java/utils/RandomGenerator.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+
+public class RandomGenerator {
+    public static List<Integer> generateNumbers(int min, int max, int count) {
```
랜덤하게 로또번호를 생성할때 조건은 1~45의 숫자를 가진 로또번호는 불변합니다. 
로또번호에 대한 검증 로직과 캐싱 기능을 사용하면 더 유연한 설계가 가능합니다.

cyrojyro replied:
LottoBall 이라는 숫자 하나를 나타내는 클래스가 추가되었습니다.

adeldel commented on src/main/java/domain/Lottos.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.List;
+
+public class Lottos {
+    private final List<Lotto> lottos;
```
미리 엠티리스트로 초기화 해주면 생성자나  매번 초기화 할 필요가 없어 보입니다.

cyrojyro replied:
반영되었습니다 👍

adeldel commented on src/main/java/view/Input.java:
```diff
@@ -0,0 +1,66 @@
+            previousBonus = scanner.nextInt();
+            scanner.nextLine();
+        } catch (Exception e) {
+            throw new InputMismatchException(Phrase.ILLEGAL_INPUT);
```
꼼꼼한 에러처리 👍 👍

adeldel commented on src/main/java/domain/LottoStatistics.java:
```diff
@@ -0,0 +1,57 @@
+                    lottoRank.getReward()
+                            .multiply(rankNumbers.get(lottoRank)));
+        }
+        return totalReward;
```
리워드의 총합을 구하고자 한다면 reduce 기능을 이용해도 좋을 것 같아요. 
```suggestion
        return rankNumbers.keySet()
                .stream()
                .map(reward -> reward.getReward().multiply(rankNumbers.get(reward)))
                .reduce(BigInteger.valueOf(0), BigInteger::add);

cyrojyro replied:
반영되었습니다 👍

adeldel reviewed (changes requested):
안녕하세요 Jack 🙌  로또 미션 구현하신 것 잘 보았습니다. 
기능 요구사항에 대해 상세히 정리 및 전반적으로 도메인 객체를 잘 분리해주셨고, 구현도 아주 잘 해주셨습니다.
구조적으로 크게 변경할 것은 없어보이지만, 조금만 수정한다면 더 좋은 코드가 될 것으로 보여 코멘트를 남겨드렸어요.
제가 드린 의견을 모두 반영하실 필요는 없으시지만,  참고해서 리팩토링을 해주십사 재요청 드립니다.
오늘도 화이팅하세요 💪

cyrojyro commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,70 @@
+    public static final int LOTTO_PRICE = 1000;
+    public static final int LOTTO_COUNT = 6;
+    public static final int LOTTO_MIN = 1;
+    public static final int LOTTO_MAX = 45;
```
아래 3개는 protected 변경하였습니다. LOTTO_PRICE는 컨트롤러에서도 사용하고 있어 public으로 해야할것 같습니다.

cyrojyro commented on src/main/java/text/Phrase.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.math.BigInteger;
+
+public class Phrase {
```
의미가 더 명확해보이게 Text 클래스로 변경하였습니다.

cyrojyro commented on src/main/java/domain/LottoRank.java:
```diff
@@ -0,0 +1,74 @@
+    private final List<Integer> validSameBalls;
+    private final List<Boolean> validBonusBall;
+    private final BigInteger reward;
+    private final String description;
```
1. 반영되었습니다.
2. 보너스번호 포함이 (필요하다, 필요하지 않다) 2가지 state로 구분하면 LottoRank 구하는 시점에 메서드에서 2, 3등과 나머지를 다르게 취급하여 분기해야합니다.
이런 2, 3등에 대한 하드코딩을 피하기 위해 (필요하다, 필요하지 않다, 무관하다) 3가지 state를 나타내는 enum을 사용했습니다.
3, 4. 반영되었습니다. 🥇

adeldel commented on src/main/java/domain/LottoBalls.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.stream.Stream;
+
+public class LottoBalls {
+    private final List<LottoBall> LottoBalls;
```
멤버변수명은 소문자로 시작하는게 관례입니다.

adeldel commented on src/main/java/domain/LottoRank.java:
```diff
@@ -0,0 +1,76 @@
+                .filter(lottoRank -> lottoRank.checkRank(count, validBonusBall))
+                .findFirst();
+
+        return result.orElse(LottoRank.NONE);
```
굳이 나눠서 할 필요 없이 합칠수 있을 것 같습니다.
```suggestion
        return Arrays.stream(LottoRank.values())
                .filter(lottoRank -> lottoRank.checkRank(count, validBonusBall))
                .findFirst().orElse(LottoRank.NONE);return Arrays.stream(LottoRank.values())
                .filter(lottoRank -> lottoRank.checkRank(count, validBonusBall))
                .findFirst()
                .orElse(LottoRank.NONE);

adeldel commented on src/main/java/domain/LottoBalls.java:
```diff
@@ -0,0 +1,58 @@
+        this.LottoBalls = LottoBalls;
+    }
+
+    public static List<LottoBall> asList(Integer... args) {
```
테스트에서만 사용하는 정적메소드인듯  합니다. 테스트에서 공통으로 사용하고 싶다면 도메인 객첵가 아닌 테스트패키지 하위에 공통으로 사용할 클래스를 만들어 사용하셔도 됩니다.

adeldel commented on src/main/java/domain/LottoBalls.java:
```diff
@@ -0,0 +1,58 @@
+
+    public int size() {
+        return LottoBalls.size();
+    }
```
stream() 이나 size() 메소드도 너무 쪼개놓으신거 아닌가요? 😅  이런 로직은 그대로 사용처에서 `LottoBalls` 를 얻어서 사용하시면 됩니다.

adeldel reviewed (approved):
잭 안녕하세요. 리뷰 요청사항에 대해 의견 반영을 잘 해주셔서 감사합니다.
추가 코멘트 드린 부분은 참고하시어,  다음 스텝에 반영해주시면 됩니다.  
( 제가 드린 코멘트가 정답은 아니므로, 다른 의견이 있으시면 코멘트 남겨주시고 잭이 맞다고 생각하시는 방향으로 하셔도 됩니다. 😄 )
그럼 다음 진행을 위해 머지해둘께요. 
수고 많으셨습니다. 🎉 🎉 🎉

owenoweny committed:
feat: 로또 클래스 작성

- 랜덤 로또 발행 로직, 단위 테스트 구현

owenoweny committed:
docs: 기능 목록 작성

owenoweny committed:
feat: 로또 발행 기능 구현

owenoweny committed:
feat: 구입 금액에 따른 로또 발행 로직 구현

- 도메인 클래스 패키지 이동
- Lottos.java, LottoMachine.java 구현

owenoweny committed:
feat: 로또 당첨 결과 계산 기능 구현

owenoweny committed:
feat: 수익률 계산 기능 구현

owenoweny committed:
feat: 로또 당첨 결과 반환 기능 구현

- WinningResult의 팩토리 메소드명 변경

owenoweny committed:
fix: WinningResult - 당첨되지 않은 경우를 확인하지 못하는 버그 수정

owenoweny committed:
refactor: 불필요한 클래스 삭제

owenoweny committed:
refactor: 로또 번호 클래스로 분리

owenoweny committed:
refactor: 불필요한 메소드 삭제

owenoweny committed:
refactor: LottoMachine의 역할 분리

- LottoMachine -> LottoManager로 클래스명 변경
- LottoMachine: 금액을 입력받아 로또리스트 객체를 반환하는 팩토리 클래스 구현

owenoweny committed:
refactor: 불필요한 메소드, 테스트 삭제

owenoweny committed:
feat: 콘솔 진행 controller, view 구현

owenoweny committed:
refactor: 패키지 구조 수정

owenoweny committed:
refactor: 클린 코드, 컨벤션 리팩토링

- 불필요한 오버라이딩 삭제
- 메소드 순서 변경
- 불필요한 개행 삭제
- 메소드명 수정
- 메소드 라인 줄임

owenoweny committed:
test: 불필요한 @DisplayName 삭제, 테스트 메소드 삭제

owenoweny committed:
docs: README 수정

sah3122 reviewed (changes requested):
안녕하세요. 로또 미션 리뷰를 맡게된 이동철 입니다 😄 
로또 미션 구현 잘해주셨습니다 👍 
몇가지 코멘트 남겨두었는데 확인 부탁드려요 🙏

owenoweny committed:
refactor: 투입 금액을 나타내는 클래스를 생성해 역할 분리

owenoweny committed:
refactor: Controller 가독성을 늘리기 위한 메소드 분리

owenoweny committed:
refactor: 유틸성 클래스 객체가 생성되지 않도록 수정

owenoweny committed:
feat: WinningLotto - 보너스 번호와 로또 번호의 중복 검증 로직 구현

owenoweny committed:
refactor: Lottos.compare rename

owenoweny committed:
test: 도메인 로직 테스트 코드 추가

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
남겨주신 질문에 대해 코멘트 추가해두었는데 확인 부탁드려요 ! 
다음 미션도 기대하겠습니다 😄
